### PR TITLE
Initial commit for CRR in Java

### PR DESCRIPTION
**Pull request by AdelGhabboun**
## Purpose

* ...

## Does this introduce a breaking change?

```
[ ] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
Comment (paulomarquesc) [inline .idea/.gitignore:1]: What this .idea folder is? Is it the folder used by the IDE you're developing this? If this is the case, can you please add the .idea folder under the main .gitignore file and remove the folder from the repo? This comment is for every entry we see below that is related to anything inside .idea

File: README.md [hunk 2]
```diff
@@ -1,57 +1,113 @@
...
-- Link to similar sample
-- ...
+---
+page_type: sample
+languages:
+- Java
+products:
+- azure
+- azure-netapp-files
+description: "This project demonstrates how to use Java with NetApp Files SDK for Microsoft.NetApp resource provider to deploy a cross-region replication for NFS 4.1 Volume."
```
Comment (paulomarquesc) [inline README.md:8]:
```suggestion
description: "This project demonstrates how to use Java with NetApp Files SDK for Microsoft.NetApp resource provider to enable cross-region replication on NFS 4.1 Volume."
```

File: README.md [hunk 3]
```diff
@@ -1,57 +1,113 @@
...
+---
+
+# Azure NetAppFiles Cross-Region Replication (CRR) SDK NFS 4.1 Sample Java
+
+This project demonstrates how to deploy a cross-region replication with enabled with NFS 4.1 protocol volume using Java and Azure NetApp Files SDK.
```
Comment (paulomarquesc) [inline README.md:13]:
```suggestion
This project demonstrates how to deploy a volume based on NFSv4.1 protocol with cross-region replication enabled using Java and Azure NetApp Files SDK.
```

File: README.md [hunk 4]
```diff
@@ -1,57 +1,113 @@
...
+	| Primary Capacity pool 
+		| Primary NFS v4.1 Volume 
+		
+ * Secondary ANF Account
+	| Secondary Capacity pool
+		| Secondary NFS v.1 Data Replication Volume with referencing to the primary volume Resource ID
+			
+ * Authorize Source volume with Desitnation Volume Resource ID
+ 
+ * Finally, the clean up process takes place (not enabled by default, please change the variable shouldCleanUp to true at main file if you want the clean up code to take a place),deleting all resources in the reverse order following the hierarchy otherwise we can't remove resources that have nested resources still live. You will also notice that the clean up process uses a function called WaitForNoANFResource, at this moment this is required so we can workaround a current ARM behavior of reporting that the object was deleted when in fact its deletion is still in progress.
```
Comment (paulomarquesc) [inline README.md:28]: This should be moved to the "What is anf-java... " section. You can add a simple "* Cleanup resources if enabled"

File: README.md [hunk 5]
```diff
@@ -1,57 +1,113 @@
...
+       ```
+        Bash
+        ```bash
+        export AZURE_AUTH_LOCATION=/sdksamples/azureauth.json
+        ``` 
+
+
+# What is anf-java-crr-sdk-nfs4.1-sample.dll doing? 
+
+This sample project is dedicated to demonstrate how to enable cross-region replication in Azure NetApp Files for a NFS v4.1 enabled volume, similar to other examples, the authentication method is based on a service principal, this project will create two ANF Accounts in different regions with capacity pool. A single volume using Premium service level tier in the Source ANF, and Data Replication Volume with Standard service level tier in the destination region.
```
Comment (paulomarquesc) [inline README.md:69]:
```suggestion
This sample project is dedicated to demonstrate how to enable cross-region replication in Azure NetApp Files for an NFS v4.1 enabled volume, similar to other examples, the authentication method is based on a service principal, this project will create two ANF Accounts in different regions with capacity pool. A single volume using Premium service level tier in the Source ANF, and Data Replication Volume with Standard service level tier in the destination region. 
```

File: README.md [hunk 6]
```diff
@@ -1,57 +1,113 @@
...
+
+| Folder         | FileName                    | Description                                                                                                                                                                                                                                                               |
+|----------------|-----------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Root\\^           | main.java                   | Reads configuration, authenticates, executes all operations
+| Root\\^           | Creation.java               | Class that contains all resource creation loops, following the hierarchy logic in order to successfully deploy Azure NetApp Files resources
+| Root\\^           | Cleanup.java                | Class that performs cleanup of all artifacts that were created during this sample application. Its call is commented out by default in main.java
```
Comment (paulomarquesc) [inline README.md:79]: The description does not match the new way to do by using the cleanup variable.

File: README.md [hunk 7]
```diff
@@ -1,57 +1,113 @@
...
+| Root\\^\common    | CommonSdk.java              | Class dedicated to common operations related to ANF's SDK
+| Root\\^\common    | ResourceUriUtils.java       | Class that exposes a few methods that help parsing Uri's, building new Uri's, or getting a resource name from a Uri, etc
+| Root\\^\common    | ServiceCredentialsAuth.java | A small support class for extracting and creating credentials from a File
+| Root\\^\common    | Utils.java                  | Class that exposes methods that help with getting the configuration object, byte conversion, etc
+| Root\\^\model     | AzureAuthInfo.java          | POJO class to hold authentication information
```
Comment (paulomarquesc) [inline README.md:84]: What is a POJO Class?

File: README.md [hunk 8]
```diff
@@ -1,57 +1,113 @@
...
+ * Authorize Source volume with Desitnation Volume Resource ID
+ 
+ * Finally, the clean up process takes place (not enabled by default, please change the variable shouldCleanUp to true at main file if you want the clean up code to take a place),deleting all resources in the reverse order following the hierarchy otherwise we can't remove resources that have nested resources still live. You will also notice that the clean up process uses a function called WaitForNoANFResource, at this moment this is required so we can workaround a current ARM behavior of reporting that the object was deleted when in fact its deletion is still in progress.
+
+
+If you don't already have a Microsoft Azure subscription, you can get a FREE trial account [here](http://go.microsoft.com/fwlink/?LinkId=330212).
+
+## Prerequisites
+
+1. Azure Subscription
```
Comment (paulomarquesc) [inline README.md:35]: I remember that you had to install mvn in your machine to make this run, if that is the case, we need to add it as a pre-req

File: README.md [hunk 9]
```diff
@@ -1,57 +1,113 @@
...
+    ```
+1. Run the console application
+    ```powershell
+    mvn exec:java -Dexec.mainClass="sdk.sample.main"
+
+Sample output
+![e2e execution](./media/e2e-execution.png)
+
+# References
+
```
Comment (paulomarquesc) [inline README.md:108]: I think we have documentation available for CRR, can you please reference that here?

File: src/main/Cleanup.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+public class Cleanup {
+    /**
+     * Deletes destination replication connection
+     * @param anfClient Azure NetApp Files Management Client
+     * @param destinationResourceGroupName Resource Group name of the destination ANF
+     * @param destinationAnfAccountName Destination ANF account name
+     * @param destinationCapacityPoolName Destination ANF capacity pool name
+     * @param destinationVolumeName Destination ANF volume name
+     */
+    public static CompletableFuture<Observable<Void>> deleteReplicationOnDestinationAsync(AzureNetAppFilesManagementClientImpl anfClient, String destinationResourceGroupName, String destinationAnfAccountName, String destinationCapacityPoolName, String destinationVolumeName)
```
Comment (paulomarquesc) [inline src/main/Cleanup.java]: I think this method should be named more generically, following all arguments and variables, if we can define all these settings, e.g. we don't need to say destinationXYZ,

File: src/main/Creation.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    {
+        List<ExportPolicyRule> ruleList = new ArrayList<>();
+        ruleList.add(new ExportPolicyRule()
+        .withAllowedClients("0.0.0.0")
+        .withRuleIndex(1)
+        .withUnixReadWrite(true)
+        .withUnixReadOnly(false)
+        .withCifs(false)
+        .withNfsv3(false)
+        .withNfsv41(true));
```
Comment (paulomarquesc) [inline src/main/Creation.java]: Just for the sake of visualization, can we add a tab here?

File: src/main/Creation.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        volumeInner.withUsageThreshold(volumeSize);
+        volumeInner.withProtocolTypes(protocols);
+        volumeInner.withLocation(location);
+        volumeInner.withDataProtection(new VolumePropertiesDataProtection()
+        .withReplication(new ReplicationObject()
+        .withEndpointType(EndpointType.DST)
+        .withRemoteVolumeRegion(primaryVolumeLocation)
+        .withRemoteVolumeResourceId(primaryVolumeId)
+        .withReplicationSchedule(ReplicationSchedule.HOURLY)));
+
```
Comment (paulomarquesc) [inline src/main/Creation.java]: Include a tab

File: src/main/Creation.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        protocols.add("NFSv4.1");
+
+        VolumeInner volumeInner = new VolumeInner();
+        volumeInner.withCreationToken(volumeName);
+        volumeInner.withExportPolicy(exportPolicy);
+        volumeInner.withServiceLevel(ServiceLevel.fromString(serviceLevel));
+        volumeInner.withSubnetId(subnetId);
+        volumeInner.withUsageThreshold(volumeSize);
+        volumeInner.withProtocolTypes(protocols);
+        volumeInner.withLocation(location);
```
Comment (paulomarquesc) [inline src/main/Creation.java]: Can we use the same pattern as in line 68?

File: src/main/Creation.java [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        volumeInner.withDataProtection(new VolumePropertiesDataProtection()
+        .withReplication(new ReplicationObject()
+        .withEndpointType(EndpointType.DST)
+        .withRemoteVolumeRegion(primaryVolumeLocation)
+        .withRemoteVolumeResourceId(primaryVolumeId)
+        .withReplicationSchedule(ReplicationSchedule.HOURLY)));
```
Comment (paulomarquesc) [inline src/main/Creation.java]: Same case of pattern, can we use the same as in line 68?

File: src/out/production/netappfiles-java-crr-sdk-sample/.idea/.gitignore [hunk 15]
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
Comment (paulomarquesc) [inline src/out/production/netappfiles-java-crr-sdk-sample/.idea/.gitignore:1]: I guess, this /src/out folder is the locally compiled output, correct? If that is the case, lets add this folder as well to the the main .gitignore file.

**Review by paulomarquesc — CHANGES_REQUESTED** [review]
I left a few comments, plus there is this target folder that contains binaries, are we supposed to provide those? I guess this is also product of running mvn locally, if that is the case we should add the target folder to .gitignore file. If that is not the case, then we must ask customer to go to the official download page of the binaries since we may not be authorized to redistribute binary files from our repo.

File: src/main/Creation.java (hunk 13 above)
Comment (AdelGhabboun) [reply src/main/Creation.java]: I have done it but for a reason I don't know Java needs to cast to the same type.

File: .idea/.gitignore (hunk 1 above)
Comment (AdelGhabboun) [reply .idea/.gitignore:1]: yes it is added by the IDE

File: README.md (hunk 7 above)
Comment (AdelGhabboun) [reply README.md:82]: Plain old Java object, equivalent to POCO in .NET

**Pull request merged**